File: caller.py
import sys
import chalk
import json
from Naked.toolshed.shell import muterun_js

def drinks(dictionary):
  output = []
  for drink in dictionary['drinks']:
    output.append(drink['strDrink'])
  return output

response = muterun_js('./callable.js', 'search margarita')
if response.exitcode == 0:
  try:
    output = response.stdout.decode("utf-8")
    print(chalk.blue('raw output from JS'), output)
    dictionary = json.loads(output)
    print(chalk.green('decoded output as Python dictionary'), dictionary)
    print(chalk.green('assert Python dictionary keys can be read'), drinks(dictionary))
  except Exception as e:
    print(chalk.red(e))
else:
  sys.stderr.write(chalk.red(str(response.stderr)))

File: caller.js
const chalk = require('chalk')
const spawn = require('child_process').spawn
console.log(chalk.yellow('requesting...'))
const search = spawn('python',['./callable.py', 'search', 'margarita'])
let buffer = ''
const onData = data => {
  console.log(chalk.yellow('buffering...'))
  buffer += data
}
const onEnd = () => {
  console.log(chalk.yellow('buffer complete !'))
  try {
    console.log(chalk.blue('raw output from Python'), buffer)
    const js = JSON.parse(buffer)
    console.log(chalk.green('raw output as JS object'), js)
    console.log(chalk.green('assert JS object properties can be read'), js.drinks.map(({ strDrink }) => strDrink))
  } catch (e) {
    console.error(chalk.red('there was an error'), buffer)
  }
}
search.stdout.on('data', onData)
search.stdout.on('end', onEnd)
